tomtom-navigation-core-examples/usecase/src/main/kotlin/com/example/usecase/MainViewModel.kt

```diff
@@ -0,0 +1,264 @@
+            navigationVisualization.addRouteClickedListener(routeClickListener)
+        }
+    }
+
+    override fun onLocationUpdate(location: GeoLocation) {
+        _location.value = location
+        mapLocationProvider.removeOnLocationUpdateListener(this)
+    }
+
+    fun registerLocationUpdatesListener() {
+        mapLocationProvider.addOnLocationUpdateListener(this)
+    }
+
+    /**
+     * Used to start navigation based on a tapped route, if navigation is not already running.
+     * - Then start the navigation using the selected route.
+     */
+    private val routeClickListener = RouteClickedListener { routingRoute, _ ->
+        if (!isNavigationRunning()) {
+            startNavigation(routingRoute)
+        }
+    }
+
+    private fun startNavigation(route: Route) {
+        setSimulationLocationProviderToNavigation(route)
+        setUpMapMatchedLocationProvider()
+        _requestNavigationStart.value = NavigationStartParameters(
+            routePlan = com.tomtom.sdk.navigation.RoutePlan(
+                route = route,
+                routePlanningOptions = checkNotNull(routePlanningOptions) {
+                    "RoutePlanningOptions is not initialized"
+                }),
+            tomTomNavigation = tomTomNavigation,
+            locationProvider = mapMatchedLocationProvider
+        )
+    }
+
+    /**
+     * Used to calculate a route using the following parameters:
+     * - InstructionPhoneticsType - This specifies whether to include phonetic transcriptions in the response.
+     * - ExtendedSections - This specifies whether to include extended guidance sections in the response, such as sections of type road shield, lane, and speed limit.
+     */
+    fun calculateRoute(origin: GeoPoint, destination: GeoPoint) {
+        val itinerary = Itinerary(origin = origin, destination = destination)
+        routePlanningOptions = RoutePlanningOptions(
+            itinerary = itinerary,
+            guidanceOptions = GuidanceOptions(
+                phoneticsType = InstructionPhoneticsType.Ipa,
+                extendedSections = ExtendedSections.All
+            ),
+            vehicle = Vehicle.Car()
+        )
+        routePlanner.planRoute(routePlanningOptions!!, routePlanningCallback)
+    }
+
+    /**
+     * The RoutePlanningCallback itself has three methods.
+     * - The `onFailure()` method is triggered if the request fails.
+     * - The `onSuccess()` method returns RoutePlanningResponse containing the routing results.
+     * - The `onRoutePlanned()` method is triggered when each route is successfully calculated.
+     *
+     * This example draws the first retrieved route on the map.
+     * You can show the overview of the added routes using the TomTomMap.zoomToRoutes(Int) method.
+     * Note that its padding parameter is expressed in pixels.
+     */
+    private val routePlanningCallback = object : RoutePlanningCallback {
+        override fun onSuccess(result: RoutePlanningResponse) {
```

Comment:
The APIs in Navigation Visualization are made to be called from the UI thread. So im not sure if this callback will be executed there. But if it is the case ignore my comment.

Reply:
The callback is executed on the main thread